### PR TITLE
Refactor

### PR DESCRIPTION
**Pull request by dorssel**
- gets rid of `Void` kludge (and other generic oddities) for the non-generic debouncer
- supports both `Trigger()` and `Trigger(data)` for generic debouncer, where `Trigger()` is still lockless
- adds `Reset` with data for generic debouncer
- resolves race condition
- rename to `DataLimit`
- adds `InvalidOperationException` when triggering with data beyond the limit while a handler is already running
- abstracts to a `DebouncerBase` class with a trivial non-generic `Debouncer` derived
- minimizes the generic `Debouncer` derived to just that what it actually adds

### CODE REVIEW

**Comment by dorssel** [comment]
Needs some more unit tests and better documentation. But code is ready.

**Comment by dorssel** [comment]
@amvandenberg 
BTW, even though we're not using it anymore, there already was https://learn.microsoft.com/en-us/dotnet/api/system.void?view=netstandard-1.2

File: Debounce/Generic/IDebouncer.cs [hunk 1]
```diff
@@ -10,49 +10,24 @@ namespace Dorssel.Utilities.Generic;
...
-    /// <exception cref="ObjectDisposedException">The object has been disposed.</exception>
-    public TimeSpan TimingGranularity { get; set; }
-
-    /// <summary>
-    /// Gets or sets the minimum number of calls to <see cref="Trigger"/> after which a new Debounced event will fire.
-    /// </summary>
-    /// <exception cref="ObjectDisposedException">The object has been disposed.</exception>
-    /// <exception cref="ArgumentOutOfRangeException">The value is less than 1.</exception>
-    public int DebounceAfterTriggerCount { get; set; }
+    public long Reset(out IReadOnlyList<TData> data);
```
Comment (amvandenberg) [inline Debounce/Generic/IDebouncer.cs:32]: Missing <param> documentation for `data`?

File: UnitTests/Generic/DebouncedEventArgsGenericTests.cs [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+        IEnumerator IEnumerable.GetEnumerator() => new MockEnumerator(this);
+
+        public IEnumerator<int> GetEnumerator() => new MockEnumerator(this);
+    }
+
+    public static IEnumerable<object[]> ValidCounts
+    {
+        get
+        {
+            yield return new object[] { 1L, new MockReadOnlyList(0) };
```
Comment (amvandenberg) [inline UnitTests/Generic/DebouncedEventArgsGenericTests.cs:66]: The documentation seems to say that an empty `triggerData` would throw an `ArgumentOutOfRangeException`?

File: Debounce/Generic/DebouncedEventArgs.cs [hunk 3]
```diff
@@ -7,60 +7,57 @@
...
-    /// <param name="boundsCheck">If <c>true</c>, <paramref name="count"/> is checked to be within its valid range.</param>
+    /// <param name="count"></param>
     /// <param name="triggerData">
-    /// Accumulated data from each individual trigger, or empty when buffering is disabled in the <see cref="Debouncer{TData}"/>
+    /// Accumulated data from each individual trigger. Must not be empty if <paramref name="boundsCheck"/> is <c>true</c>.
     /// </param>
-    /// <exception cref="ArgumentOutOfRangeException">Thrown when <paramref name="boundsCheck"/> is <c>true</c> and <paramref name="count"/> is not greater than 0.</exception>
-    protected DebouncedEventArgs(long count, bool boundsCheck, IReadOnlyList<TData> triggerData)
+    /// <param name="boundsCheck">If <c>true</c>, <paramref name="triggerData"/> is checked to be non-empty.</param>
+    /// <exception cref="ArgumentOutOfRangeException">Thrown when <paramref name="boundsCheck"/> is <see langword="true" /> and <paramref name="triggerData"/> is an empty list.</exception>
```
Comment (amvandenberg) [inline Debounce/Generic/DebouncedEventArgs.cs:35]: Documentation does not match the actual check, which seems to check whether `triggerData` contains more items than `count`, not the empty list check.

File: Debounce/Generic/DebouncedEventArgs.cs [hunk 4]
```diff
@@ -7,60 +7,57 @@
...
-    /// <param name="count">The number of triggers accumulated since the previous event was sent.
-    /// <para>Must be greater than 0.</para>
-    /// </param>
+    /// <param name="count"></param>
     /// <param name="triggerData">
-    /// Accumulated data from each individual trigger, or empty when buffering is disabled in the <see cref="Debouncer{TData}"/>
+    /// Accumulated data from each individual trigger. Must not be empty."/>
     /// </param>
-    /// <exception cref="ArgumentOutOfRangeException">Thrown when <paramref name="count"/> is not greater than 0.</exception>
+    /// <exception cref="ArgumentOutOfRangeException">Thrown when <paramref name="triggerData"/> is an empty list.</exception>
```
Comment (amvandenberg) [inline Debounce/Generic/DebouncedEventArgs.cs:21]: Documentation does not match the actual check, which seems to check whether `triggerData` contains more items than `count`, not the empty list check.

File: Debounce/DebouncerBase.cs [hunk 5]
```diff
@@ -0,0 +1,412 @@
...
+    , IDisposable
+    where TEventArgs : DebouncedEventArgs
+{
+    private protected DebouncerBase()
+    {
+        Timer = new Timer(OnTimer, null, Timeout.InfiniteTimeSpan, Timeout.InfiniteTimeSpan);
+    }
+
+    /// <summary>
+    ///
```
Comment (amvandenberg) [inline Debounce/DebouncerBase.cs:24]: Missing documentation. Or to be removed...

File: Debounce/DebouncerBase.cs [hunk 6]
```diff
@@ -0,0 +1,412 @@
...
+        {
+            lock (LockObject)
+            {
+                return _Benchmark;
+            }
+        }
+    }
+
+    /// <summary>
+    ///
```
Comment (amvandenberg) [inline Debounce/DebouncerBase.cs:51]: Missing documentation. Or to be removed...

File: Debounce/DebouncerBase.cs [hunk 7]
```diff
@@ -0,0 +1,412 @@
...
+    long Count;
+    readonly Stopwatch FirstTrigger = new();
+    readonly Stopwatch LastTrigger = new();
+    readonly Stopwatch LastHandlerStarted = new();
+    readonly Stopwatch LastHandlerFinished = new();
+    readonly Timer Timer;
+    bool TimerActive;
+    private protected bool SendingEvent;
+
+    internal static long AddWithClamp(long left, long right)
```
Comment (amvandenberg) [inline Debounce/DebouncerBase.cs:64]: This method could use a bit of documentation. Although the implementation is clear, when using the function you might not want to reread the doc.

File: Debounce/Generic/Debouncer.cs [hunk 8]
```diff
@@ -5,452 +5,116 @@
...
-                    throw new ArgumentOutOfRangeException(fieldName, $"{fieldName} must not be infinite (-1 ms).");
-                }
-            }
-            else if (value < TimeSpan.Zero)
-            {
-                throw new ArgumentOutOfRangeException(fieldName, $"{fieldName} must not be negative.");
-            }
-            field = value;
-            LockedReschedule();
+            return new ReadOnlyCollection<TData>(Interlocked.Exchange(ref TriggerData, []));
```
Comment (amvandenberg) [inline Debounce/Generic/Debouncer.cs:103]: Nice usage of the `Interlocked.Exchange()`! although you're already locked so no specific need for an atomic operation, right?

**Review by amvandenberg — COMMENTED** [review]
Very nice and clean!

File: Debounce/Generic/IDebouncer.cs (hunk 1 above)
Comment (dorssel) [reply Debounce/Generic/IDebouncer.cs:33]: Fixed

File: UnitTests/Generic/DebouncedEventArgsGenericTests.cs (hunk 2 above)
Comment (dorssel) [reply UnitTests/Generic/DebouncedEventArgsGenericTests.cs:66]: Fixed the documentation.

File: Debounce/DebouncerBase.cs (hunk 7 above)
Comment (dorssel) [reply Debounce/DebouncerBase.cs:64]: Done

File: Debounce/DebouncerBase.cs (hunk 6 above)
Comment (dorssel) [reply Debounce/DebouncerBase.cs]: Done

File: Debounce/DebouncerBase.cs (hunk 5 above)
Comment (dorssel) [reply Debounce/DebouncerBase.cs]: Done

File: Debounce/Generic/DebouncedEventArgs.cs (hunk 4 above)
Comment (dorssel) [reply Debounce/Generic/DebouncedEventArgs.cs]: Fixed

**Comment by dorssel** [comment]
Hmmm, GitHub gets really confused after rebasing. I couldn't reply to some comments within the review window. Anyways... I resolved everything, I think. Can you have look at the latest commit?

**Review by amvandenberg — APPROVED** [review]

**Pull request merged**